src/cdk/scrolling/virtual-scroll-viewport.ts

```diff
@@ -0,0 +1,191 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Range} from '@angular/cdk/collections';
+import {CdkVirtualForOf} from '@angular/cdk/scrolling/virtual-for-of';
+import {
+  ChangeDetectionStrategy,
+  ChangeDetectorRef,
+  Component,
+  DoCheck,
+  ElementRef,
+  Inject,
+  NgZone,
+  OnDestroy,
+  OnInit,
+  ViewChild,
+  ViewEncapsulation,
+  Input,
+} from '@angular/core';
+import {fromEvent} from 'rxjs/observable/fromEvent';
+import {takeUntil} from 'rxjs/operators/takeUntil';
+import {Subject} from 'rxjs/Subject';
+import {Observable} from 'rxjs/Observable';
+import {VIRTUAL_SCROLL_STRATEGY, VirtualScrollStrategy} from './virtual-scroll-strategy';
+
+
+/** A viewport that virtualizes it's scrolling with the help of `CdkVirtualForOf`. */
+@Component({
+  moduleId: module.id,
+  selector: 'cdk-virtual-scroll-viewport',
+  templateUrl: 'virtual-scroll-viewport.html',
+  styleUrls: ['virtual-scroll-viewport.css'],
+  host: {
+    'class': 'cdk-virtual-scroll-viewport',
+  },
+  encapsulation: ViewEncapsulation.None,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  preserveWhitespaces: false,
+})
+export class CdkVirtualScrollViewport implements OnInit, DoCheck, OnDestroy {
+  private _disconnectSubject = new Subject<void>();
+
+  private _renderedRangeSubject = new Subject<Range>();
+
+  /** The direction the viewport scrolls. */
+  @Input() orientation: 'horizontal' | 'vertical' = 'vertical';
+
+  /** The element that wraps the rendered content. */
+  @ViewChild('contentWrapper') _contentWrapper: ElementRef;
+
+  /** The total size of all content, including content that is not currently rendered. */
+  get totalContentSize() { return this._totalContentSize; }
+  set totalContentSize(size: number) {
+    if (this._totalContentSize != size) {
+      this._ngZone.run(() => {
+        this._totalContentSize = size;
+        this._changeDetectorRef.markForCheck();
+      });
+    }
+  }
+  private _totalContentSize = 0;
+
+  /** The currently rendered range of indices. */
+  get renderedRange() { return this._renderedRange; }
+  set renderedRange(range: Range) {
+    if (!this._rangesEqual(this._renderedRange, range)) {
+      this._ngZone.run(() => {
+        this._renderedRangeSubject.next(this._renderedRange = range);
+        this._changeDetectorRef.markForCheck();
+      });
+    }
+  }
+  private _renderedRange: Range = {start: 0, end: 0};
+
+  /** The offset of the rendered portion of the data from the start. */
+  get renderedContentOffset(): number { return this._renderedContentOffset; }
+  set renderedContentOffset(offset: number) {
+    if (this._renderedContentOffset != offset) {
+      this._ngZone.run(() => {
+        this._renderedContentOffset = offset;
+        this._renderedContentTransform = this.orientation === 'horizontal' ?
```

Comment:
Might want to do browser prefix test here.

Reply:
We only have to support back to IE11 (+ latest 2 versions of other browsers) so should be ok